File: patterns/blink48ports.go
package patterns

import (
	"time"

	"github.com/anderstorpsfestivalen/slisko/pkg/chassi"
	"github.com/anderstorpsfestivalen/slisko/pkg/configuration"
	"github.com/anderstorpsfestivalen/slisko/pkg/traffic"
)

type Blink48Ports struct {
	ports []PortState
	style BlinkStyle
}

func (p *Blink48Ports) Render(info RenderInfo, c *chassi.Chassi) {
	for i := range p.ports {
		p.ports[i].Render()
	}
}

func (p *Blink48Ports) Info() PatternInfo {
	return PatternInfo{
		Name:     "blink48ports",
		Category: "link",
	}
}

func (p *Blink48Ports) Bootstrap(c *chassi.Chassi) {
	p.style = Cisco7609Style()

	// Get or create default traffic shaper
	shaper := GetTrafficShaper()
	if shaper == nil {
		// Fallback to default if not set
		shaper = traffic.NewShaper(configuration.DefaultTrafficShaper())
	}

	for _, card := range c.GetCardOfType("6478") {
		for _, port := range card.Link {
			time.Sleep(2 * time.Millisecond)
			p.ports = append(p.ports, p.style.CreatePort(port, shaper))
		}
	}
}

File: engine/src/patterns/cisco7609/blink_48_ports.rs
use alloc::vec::Vec;

use super::blinkstyle::cisco7609_style;
use crate::chassi::Chassi;
use crate::pattern::{BootstrapCtx, LinkActivation, Pattern, PatternInfo, RenderInfo};
use crate::patterns::blinkstyle::PortState;

/// Cisco 7609 6478 linecard ports.
#[derive(Default)]
pub struct Blink48Ports {
    ports: Vec<PortState>,
}

impl Pattern for Blink48Ports {
    fn render(&mut self, info: &RenderInfo, c: &mut Chassi) {
        for port in &mut self.ports {
            port.render(&mut c.leds, info);
        }
    }

    fn info(&self) -> PatternInfo {
        PatternInfo {
            name: "blink48ports",
            category: "link",
        }
    }

    fn bootstrap(&mut self, c: &Chassi, ctx: &mut BootstrapCtx) {
        let style = cisco7609_style();
        for index in c.link_indices_of_type("6478") {
            self.ports.push(style.create_port(index, ctx));
        }
    }

    fn restart_link_traffic(&mut self, now_ms: u32, activations: &[LinkActivation]) {
        for port in &mut self.ports {
            port.restart_link_traffic(now_ms, activations);
        }
    }
}
